netty/src/main/java/io/grpc/netty/GrpcHttp2HeadersDecoder.java

```diff
@@ -0,0 +1,498 @@
+    private static final AsciiString SCHEME_HEADER = AsciiString.of(":scheme");
+
+    private AsciiString path;
+    private AsciiString authority;
+    private AsciiString method;
+    private AsciiString scheme;
+    private AsciiString te;
+
+    GrpcHttp2RequestHeaders(int numHeadersGuess) {
+      super(numHeadersGuess);
+    }
+
+    @Override
+    public Http2Headers add(CharSequence csName, CharSequence csValue) {
+      final AsciiString name = validateName(requireAsciiString(csName));
+      final AsciiString value = requireAsciiString(csValue);
+      if (isPseudoHeader(name)) {
+        addPseudoHeader(name, value);
+        return this;
+      }
+      if (equals(TE_HEADER, name)) {
+        te = value;
+        return this;
+      }
+      return add(name, value);
+    }
+
+    @Override
+    public CharSequence get(CharSequence csName) {
+      final AsciiString name = requireAsciiString(csName);
+      if (isPseudoHeader(name)) {
+        // This code should never be reached.
+        throw new IllegalArgumentException("Use direct accessor methods for pseudo headers.");
+      }
+      if (equals(TE_HEADER, name)) {
+        return te;
+      }
+      return get(name);
+    }
+
+    private void addPseudoHeader(CharSequence csName, CharSequence csValue) {
+      AsciiString name = requireAsciiString(csName);
+      AsciiString value = requireAsciiString(csValue);
+
+      if (equals(PATH_HEADER, name)) {
+        path = value;
+      } else if (equals(AUTHORITY_HEADER, name)) {
+        authority = value;
+      } else if (equals(METHOD_HEADER, name)) {
+        method = value;
+      } else if (equals(SCHEME_HEADER, name)) {
+        scheme = value;
+      } else {
+        PlatformDependent.throwException(
+            connectionError(PROTOCOL_ERROR, "Illegal pseudo-header '%s' in request.", name));
+      }
+    }
+
+    @Override
+    public CharSequence path() {
+      return path;
+    }
+
+    @Override
+    public CharSequence authority() {
+      return authority;
+    }
+
+    @Override
+    public CharSequence method() {
+      return method;
+    }
+
+    @Override
+    public CharSequence scheme() {
+      return scheme;
+    }
+
+
+    /**
+     * This method is called in tests only.
+     */
+    @Override
+    public List<CharSequence> getAll(CharSequence csName) {
+      final AsciiString name = requireAsciiString(csName);
+      if (isPseudoHeader(name)) {
+        // This code should never be reached.
+        throw new IllegalArgumentException("Use direct accessor methods for pseudo headers.");
+      }
+      if (equals(TE_HEADER, name)) {
+        return Collections.singletonList((CharSequence) te);
+      }
+      return super.getAll(csName);
+    }
+
+    /**
+     * This method is called in tests only.
+     */
+    @Override
+    public int size() {
+      int size = 0;
+      if (path != null) {
+        size++;
+      }
+      if (authority != null) {
```

Comment:
branchy code here would be bad.  Can you push up the known header size incrementing in addPseudoHeader ?